doc/source/api.rst

```diff
@@ -1,10 +1,259 @@
+
+.. autofunction:: ray.init
+
+Defining remote functions
+-------------------------
+
+Remote functions are used to create tasks. To define a remote function, the
+``@ray.remote`` decorator is placed over the function definition.
+
+The function can then be invoked with ``f.remote``. Invoking the function
+creates a **task** which will be scheduled on and executed by some worker
+process in the Ray cluster. The call will return an **object ID** (essentially a
+future) representing the eventual return value of the task.
+
+When a remote function returns, its outputs will be serialized into a string of
+bytes and stored in an object store (whichever object store the worker that
+executed the task is connected to).
+
+Note that arguments to remote functions can be values or object IDs.
+
+.. code-block:: python
+
+  @ray.remote
+  def f(x):
+    return x + 1
+
+  x_id = f.remote(0)
+  ray.get(x_id)  # 1
+
+  y_id = f.remote(x_id)
+  ray.get(y_id)  # 2
+
+If you want a remote function to return multiple object IDs, you can do that by
+passing the ``num_return_vals`` argument into the remote decorator.
+
+.. code-block:: python
+
+  @ray.remote(num_return_vals=2)
+  def f():
+    return 1, 2
+
+  x_id, y_id = f.remote()
+  ray.get(x_id)  # 1
+  ray.get(y_id)  # 2
+
 .. autofunction:: ray.remote
+
+Getting values from object IDs
+------------------------------
+
+Object IDs can be converted into objects by calling ``ray.get`` on the object
+ID. When a driver or worker process calls ``ray.get`` on an object ID, the
+corresponding object will be transferred to the driver or worker's local object
+store (from whichever object store it is in). It will then be deserialized and
+turned into a regular Python object.
+
+Note that ``ray.get`` accepts either a single object ID or a list of object IDs.
+
+.. code-block:: python
+
+  import numpy as np
+
+  @ray.remote
+  def f():
+    return {'key1': ['value']}
+
+  # Get one object ID.
+  ray.get(f.remote())  # {'key1': ['value']}
+  # Get a list of object IDs.
+  ray.get([f.remote() for _ in range(2)])  # [{'key1': ['value']}, {'key1': ['value']}]
+
+Numpy arrays
+~~~~~~~~~~~~
+
+Numpy arrays are handled more efficiently than other data types, so **use numpy
+arrays whenever possible**.
+
+Any numpy arrays that are part of the serialized object will not be copied out
+of the object store. They will remain in the object store and the resulting
+deserialized object will simply have a pointer to the relevant place in the
+object store's memory.
+
+Since objects in the object store are immutable, this means that if you want to
+mutate a numpy array that was returned by a remote function, you will have to
+first copy it.
+
+.. autofunction:: ray.get
+
+Putting objects in the object store
+-----------------------------------
+
+The primary way that objects are placed in the object store is by being returned
+by a task. However, it is also possible to directly place objects in the object
+store using ``ray.put``.
```

Comment:
I don't think it's necessary to have the first sentence. I would start with the scenario when `ray.put` is necessary, followed by your second sentence, and then an explanation that `ray.put` also returns an object ID, same as a remote function call.

Also, I wonder if we should try to deemphasize `ray.put` and encourage people to pass around task futures instead?